### PR TITLE
Hook refactoring

### PR DESCRIPTION
**Pull request by kiip1**
### Description
Refactors some 2015 code to our latest code standards and also fixed some small bugs.
Still needs testing, since all returned lists are now immutable.

### CODE REVIEW

File: src/main/java/ch/njol/skript/Skript.java [hunk 1]
```diff
@@ -550,32 +553,24 @@ public void onEnable() {
...
+						WorldGuardHook.class, WorldGuardHook::new,
+						VaultHook.class, VaultHook::new);
+
+				for (Entry<Class<? extends Hook>, Supplier<Hook>> entry : hooks.entrySet()) {
+					if (!isHookEnabled(entry.getKey()))
+						continue;
+
+					try {
+						entry.getValue().get();
+					} catch (Exception e) {
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/Skript.java:571]:
```suggestion
					} catch (Exception ex) {
```

File: src/main/java/ch/njol/skript/hooks/SimpleHook.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	}
+
+	protected Plugin plugin() {
+		return plugin;
+	}
+	
+	protected void loadClasses() {
+		try {
+			Skript.getAddonInstance().loadClasses(getClass().getPackage().getName());
+		} catch (IOException e) {
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/hooks/SimpleHook.java:70]:
```suggestion
		} catch (IOException ex) {
```

File: src/main/java/ch/njol/skript/hooks/SimpleHook.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.UnknownNullability;
+
+import java.io.IOException;
+
+@ApiStatus.Internal
+public abstract class SimpleHook implements Hook {
+
+	private static final ArgsMessage m_hooked = new ArgsMessage("hooks.hooked");
+	private static final ArgsMessage m_hook_error = new ArgsMessage("hooks.error");
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/hooks/SimpleHook.java:35]:
```suggestion
	private static final ArgsMessage HOOKED_MESSAGE = new ArgsMessage("hooks.hooked");
	private static final ArgsMessage HOOK_ERROR_MESSAGE = new ArgsMessage("hooks.error");
```

File: src/main/java/ch/njol/skript/hooks/VaultHook.java [hunk 4]
```diff
@@ -18,58 +18,66 @@
...
+	protected void loadClasses() {
+		try {
+			boolean docs = Documentation.canGenerateUnsafeDocs();
+			if (economy != null || docs)
+				Skript.getAddonInstance().loadClasses(getClass().getPackage().getName() + ".economy");
+			if (chat != null || docs)
+				Skript.getAddonInstance().loadClasses(getClass().getPackage().getName() + ".chat");
+			if (permission != null || docs)
+				Skript.getAddonInstance().loadClasses(getClass().getPackage().getName() + ".permission");
+		} catch (IOException e) {
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/hooks/VaultHook.java:70]:
```suggestion
		} catch (IOException ex) {
```

File: src/main/java/ch/njol/skript/hooks/regions/GriefPreventionHook.java [hunk 5]
```diff
@@ -18,164 +18,185 @@
...
 		try {
 			claimsField = DataStore.class.getDeclaredField("claims");
 			claimsField.setAccessible(true);
 			if (!List.class.isAssignableFrom(claimsField.getType()))
 				claimsField = null;
-		} catch (final NoSuchFieldException e) {} catch (final SecurityException e) {}
+		} catch (NoSuchFieldException | SecurityException ignored) {}
 		if (getClaim == null && claimsField == null) {
-			Skript.error("Skript " + Skript.getVersion() + " is not compatible with GriefPrevention " + plugin.getDescription().getVersion() + "."
+			Skript.error("Skript " + Skript.getVersion() + " is not compatible with GriefPrevention " + plugin().getDescription().getVersion() + "."
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/hooks/regions/GriefPreventionHook.java:74]:
```suggestion
			Skript.error("Skript " + Skript.getVersion() + " is not compatible with " name() + " " + plugin().getDescription().getVersion() + "."
```

File: src/main/java/ch/njol/skript/hooks/regions/GriefPreventionHook.java [hunk 6]
```diff
@@ -18,164 +18,185 @@
...
+
+	@Override
+	@Nullable
+	public Region getRegion_i(World world, String name) {
+		try {
+			Claim claim = getClaim(Long.parseLong(name));
+			if (claim != null && world.equals(claim.getLesserBoundaryCorner().getWorld()))
+				return new GriefPreventionRegion(claim);
+			return null;
+		} catch (NumberFormatException e) {
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/hooks/regions/GriefPreventionHook.java:107]:
```suggestion
		} catch (NumberFormatException ex) {
```

File: src/main/java/ch/njol/skript/hooks/regions/GriefPreventionHook.java [hunk 7]
```diff
@@ -18,164 +18,185 @@
...
+	Claim getClaim(long id) {
+		GriefPrevention plugin = (GriefPrevention) plugin();
+
 		if (getClaim != null) {
 			try {
 				return (Claim) getClaim.invoke(plugin.dataStore, id);
-			} catch (final IllegalAccessException e) {
-				assert false : e;
-			} catch (final IllegalArgumentException e) {
+			} catch (final IllegalAccessException | IllegalArgumentException e) {
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/hooks/regions/GriefPreventionHook.java:129]:
```suggestion
			} catch (final IllegalAccessException | IllegalArgumentException ex) {
```

File: src/main/java/ch/njol/skript/hooks/regions/GriefPreventionHook.java [hunk 8]
```diff
@@ -18,164 +18,185 @@
...
+				for (Object claim : claims) {
 					if (!(claim instanceof Claim))
 						continue;
 					if (((Claim) claim).getID() == id)
 						return (Claim) claim;
 				}
-			} catch (final IllegalArgumentException e) {
-				assert false : e;
-			} catch (final IllegalAccessException e) {
+			} catch (IllegalArgumentException | IllegalAccessException e) {
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/hooks/regions/GriefPreventionHook.java:144]:
```suggestion
			} catch (IllegalArgumentException | IllegalAccessException ex) {
```

**Review by AyhamAl-Ali — CHANGES_REQUESTED** [review]
Nice PR ⚡

A quick review on-the-fly, I will review again later probably

File: src/main/java/ch/njol/skript/Skript.java (hunk 1 above)
Comment (kiip1) [reply src/main/java/ch/njol/skript/Skript.java:571]: I disagree with all the `e` -> `ex` [Google](https://google.github.io/styleguide/javaguide.html) uses `e` too and so do we--there are 25 matches for `ex` and 100+ for `e`.

Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/Skript.java:571]: IMO `e` is misleading as it was used for event a lot, either `exception` or `ex`, just better than `e`

Comment (kiip1) [reply src/main/java/ch/njol/skript/Skript.java:571]: Events should be `event` not `e`, there's a consensus to use `e` for exceptions

Comment (kiip1) [reply src/main/java/ch/njol/skript/Skript.java:571]: If you do want to make this change there should be some discussion with multiple contributors

File: src/main/java/ch/njol/skript/hooks/regions/GriefPreventionHook.java [hunk 9]
```diff
@@ -18,164 +18,185 @@
...
 			claimsField = DataStore.class.getDeclaredField("claims");
 			claimsField.setAccessible(true);
 			if (!List.class.isAssignableFrom(claimsField.getType()))
 				claimsField = null;
-		} catch (final NoSuchFieldException e) {} catch (final SecurityException e) {}
+		} catch (NoSuchFieldException | SecurityException ignored) {}
 		if (getClaim == null && claimsField == null) {
-			Skript.error("Skript " + Skript.getVersion() + " is not compatible with GriefPrevention " + plugin.getDescription().getVersion() + "."
+			Skript.error("Skript " + Skript.getVersion() + " is not compatible with " + name() + " " + plugin().getDescription().getVersion() + "."
 					+ " Please report this at https://github.com/SkriptLang/Skript/issues/ if this error occurred after you updated GriefPrevention.");
```
Comment (bluelhf) [inline src/main/java/ch/njol/skript/hooks/regions/GriefPreventionHook.java:75]: What?

File: src/main/java/ch/njol/skript/hooks/Hook.java [hunk 10]
```diff
@@ -18,70 +18,13 @@
...
-		if (!init()) {
-			Skript.error(m_hook_error.toString(p.getName()));
-			return;
-		}
-
-		loadClasses();
-
-		if (Skript.logHigh())
-			Skript.info(m_hooked.toString(p.getName()));
+	String name();
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/hooks/Hook.java:28]: Skript has typically followed a get/set prefix style. Any particular reason for leaving it out here?

File: src/main/java/ch/njol/skript/hooks/VaultHook.java [hunk 11]
```diff
@@ -18,58 +18,66 @@
...
 		return "Vault";
 	}
-	
+
+	@Override
+	protected void loadClasses() {
+		try {
+			boolean docs = Documentation.canGenerateUnsafeDocs();
+			if (economy != null || docs)
+				Skript.getAddonInstance().loadClasses(getClass().getPackage().getName() + ".economy");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/hooks/VaultHook.java:65]: Likely worth it to use a variable for both `Skript.getAddonInstance()` and `getClass().getPackage().getName()`

File: src/main/java/ch/njol/skript/hooks/VaultHook.java [hunk 12]
```diff
@@ -18,58 +18,66 @@
...
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	@Nullable
+	private static <T> T obtain(Class<T> clazz) {
+		ServicesManager manager = Bukkit.getServicesManager();
+		RegisteredServiceProvider<T> service = manager.getRegistration(clazz);
+
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/hooks/VaultHook.java:79]:
```suggestion
		RegisteredServiceProvider<T> service = Bukkit.getServicesManager().getRegistration(clazz);
```

File: src/main/java/ch/njol/skript/hooks/regions/PreciousStonesHook.java [hunk 13]
```diff
@@ -96,50 +81,46 @@ protected Class<? extends Region> getRegionClass() {
     @YggdrasilID("PreciousStonesRegion")
     public final class PreciousStonesRegion extends Region {
 
+		@UnknownNullability
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/hooks/regions/PreciousStonesHook.java:84]:
```suggestion
	@UnknownNullability
```

File: src/main/java/ch/njol/skript/hooks/regions/PreciousStonesHook.java [hunk 14]
```diff
@@ -96,50 +81,46 @@ protected Class<? extends Region> getRegionClass() {
...
         private transient Field field;
 
-        public PreciousStonesRegion(final Field field) {
+	    @SuppressWarnings("unused")
+	    public PreciousStonesRegion() {}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/hooks/regions/PreciousStonesHook.java:88]:
```suggestion
	@SuppressWarnings("unused")
	public PreciousStonesRegion() {}
```

File: src/main/java/ch/njol/skript/hooks/regions/PreciousStonesHook.java [hunk 15]
```diff
@@ -96,50 +81,46 @@ protected Class<? extends Region> getRegionClass() {
...
         public Collection<OfflinePlayer> getMembers() {
-            @SuppressWarnings("deprecation")
-			Set<OfflinePlayer> collect = field.getAllAllowed().stream()
-                    .map(Bukkit::getOfflinePlayer)
-                    .collect(Collectors.toSet());
-            assert collect != null;
-			return collect;
+	        return field.getAllAllowed().stream()
+					.map(Bukkit::getOfflinePlayer)
+					.collect(Collectors.toSet());
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/hooks/regions/PreciousStonesHook.java:108]:
```suggestion
	        return field.getAllAllowed().stream()
				.map(Bukkit::getOfflinePlayer)
				.collect(Collectors.toSet());
```

File: src/main/java/ch/njol/skript/hooks/regions/RegionsPlugin.java [hunk 16]
```diff
@@ -18,121 +18,123 @@
...
-			public String getID(final Class<?> c) {
-				for (final RegionsPlugin<?> p : plugins)
-					if (p.getRegionClass() == c)
-						return c.getClass().getSimpleName();
+			public String getID(Class<?> type) {
+				for (RegionsPlugin plugin : plugins) {
+					if (plugin.getRegionClass() == type)
+						return type.getSimpleName();
+				}
+
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/hooks/regions/RegionsPlugin.java:51]:
```suggestion
```
same goes for other instances, i don't think a space is necessary in this case

File: src/main/java/ch/njol/skript/hooks/regions/ResidenceHook.java [hunk 17]
```diff
@@ -109,26 +101,28 @@ protected Class<? extends Region> getRegionClass() {
...
 	@YggdrasilID("ResidenceRegion")
-	public class ResidenceRegion extends Region {
-		
-		private transient ClaimedResidence res;
+	public final class ResidenceRegion extends Region {
+
+		@UnknownNullability
+		private transient ClaimedResidence residence;
+		@UnknownNullability
 		final World world;
```
[the remark this answers is not in the document]
Comment (kiip1) [reply src/main/java/ch/njol/skript/hooks/regions/ResidenceHook.java:109]: The IDE complains about residence not being initialized, but we know it's nothing to worry about.

File: src/main/java/ch/njol/skript/hooks/SimpleHook.java [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.io.IOException;
+
+@ApiStatus.Internal
+public abstract class SimpleHook implements Hook {
+
+	private static final ArgsMessage HOOKED_MESSAGE = new ArgsMessage("hooks.hooked");
+	private static final ArgsMessage HOOK_ERROR_MESSAGE = new ArgsMessage("hooks.error");
+
+	@UnknownNullability
```
[the remark this answers is not in the document]
Comment (kiip1) [reply src/main/java/ch/njol/skript/hooks/SimpleHook.java:37]: Yes, but as nothing gets executed if its null, we can expect it to be not null when used. `UnknownNullability` is used to stop the IDE from complaining.

**Review by sovdeeth — APPROVED** [review]

File: src/main/java/ch/njol/skript/hooks/SimpleHook.java [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+
+	protected P plugin() {
+		return plugin;
+	}
+
+	protected void loadClasses() {
+		try {
+			Skript.getAddonInstance().loadClasses(getClass().getPackage().getName());
+		} catch (IOException e) {
+			throw new RuntimeException(e);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/hooks/SimpleHook.java:87]:
```suggestion
			throw new RuntimeException("Failed to load hook " + getName(), e);
```
Might be worth providing a basic message too

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Some thoughts based on the history of this system. Nice work!

**Comment by sovdeeth** [comment]
Kiip has indicated that they will not be continuing their PRs, so this is open for adoption (it will be closed for inactivity in December)

**Comment by sovdeeth** [comment]
Closing due to inactivity, please re-open if you want to take it over. Open to anyone.

**Pull request closed without merging**